### PR TITLE
config update support, cluster start fix, cluster service/stop ..

### PR DESCRIPTION
**Pull request by srids**
this PR has few features which overlap and likely fix other pending PRs. more:

PR #19 : have some changes from this one here and have fixes for the stop service and config updates logic
PR #30 : needed this for our automation. so have that change here too

additionally have support for restarting components which have stale configs.

addresses my previously raised issue #33

### CODE REVIEW

**Comment by srids** [comment]
With this, PR #19 and PR #30 can be ignored/abandoned

**Comment by jimbobhickville** [comment]
@srids - I don't really maintain this any more as the product I was working on that utilized it was shut down and I've switched jobs. Would you like to be added as a collaborator to help maintain it in the future?

**Comment by srids** [comment]
sure absolutely Greg. Much thanks for getting us the head start :)

**Comment by jimbobhickville** [comment]
I've added you on github. I'll need your pypi username to add you there so you can push out releases too.

**Comment by srids** [comment]
awesome - thanks. its: srids

**Comment by jimbobhickville** [comment]
Ok, you should be set. Just remember to increment the version number before running the build and release script or pypi will complain :D

**Comment by dimaspivak** [comment]
Hey @jimbobhickville wanna add me as a maintainer, as well? I'm also sitting on a list of changes in a fork and this way @srids and I can do reviews for each other and keep your baby alive. :)

**Comment by jimbobhickville** [comment]
That ok with you @srids?

**Comment by srids** [comment]
sure @jimbobhickville that will work best

File: ambariclient/models.py [hunk 1]
```diff
@@ -738,13 +760,109 @@ def restart(self, component_names=None):
...
+        new_item = {'Clusters': {'desired_config': [{'type': self.parent.type, 'tag': new_tag }]}}
+        previous_tag = self.parent.cluster.desired_configs[self.parent.type]['tag']
+        self.inflate()
+        previous_tag_item = None
+        for model in self._models:
+            if model.tag == previous_tag:
+                previous_tag_item = (model.items[0] if (model.items and len(model.items) > 0) else None)
+                break
+
+        def uniquify(previous_tag_item, key, value):
```
Comment (dimaspivak) [inline ambariclient/models.py]: Perhaps this just be moved into a function in `utils.py`? I'm also a bit unclear as to its purpose and how it handles attributes of type `dict`, `list`, and other. Can you add some documentation around the intended use? If it simplifies things, I'd prefer to just handle one kind of type (if possible) to just keep the complexity down.

File: ambariclient/models.py [hunk 2]
```diff
@@ -738,13 +760,109 @@ def restart(self, component_names=None):
...
+        }))
+        return self
+
+
+class ConfigurationItemCollection(base.QueryableModelCollection):
+    @property
+    def url(self):
+        return self.parent.url
+
+    def create(self, tag=None, **kwargs):
```
Comment (dimaspivak) [inline ambariclient/models.py]: Can we add a little more documentation describing this function? I know Ambari has a nightmarish way of handling configuration changes, so sharing the knowledge you've clearly picked up the hard way would be useful to everyone :)

File: ambariclient/models.py [hunk 3]
```diff
@@ -1001,6 +1124,20 @@ def commission(self, service, hosts, commission_type):
         }))
         return self
 
+    def restart_stale_config_components(self, context=None):
+        """Restart all required components resulting from stale configs"""
+        self.load(self.client.post(self.cluster.requests.url, data={
+            "RequestInfo": {
+                "context": (context if context else "Restart All Required"),
```
Comment (dimaspivak) [inline ambariclient/models.py]: Can use `context or 'Restart All Required'` instead

File: ambariclient/models.py [hunk 4]
```diff
@@ -738,13 +760,109 @@ def restart(self, component_names=None):
...
+        previous_tag_item = None
+        for model in self._models:
+            if model.tag == previous_tag:
+                previous_tag_item = (model.items[0] if (model.items and len(model.items) > 0) else None)
+                break
+
+        def uniquify(previous_tag_item, key, value):
+            attrib_value = None
+            if previous_tag_item:
+                attrib_value = (previous_tag_item[key] if key in previous_tag_item else None)
```
Comment (dimaspivak) [inline ambariclient/models.py]: Can probably do `previous_tag_item.get(key)`?

File: ambariclient/models.py [hunk 5]
```diff
@@ -738,13 +760,109 @@ def restart(self, component_names=None):
...
+    def url(self):
+        return self.parent.url
+
+    def create(self, tag=None, **kwargs):
+        """This collection requires you to PUT the entire collection every time you add to it"""
+        # support for create(properties='{"hadoop.proxyuser.xyz.groups" : "*"}')
+        # a PUT to "http://localhost:8080/api/v1/clusters/cluster"
+        # in format as: [{"Clusters":{"desired_config":[{"type":"core-site", "tag":"somrandchars",
+        #                                                "properties":{"hadoop.proxyuser.xyz.groups" : "*", ... }]}}]
+        new_tag = (tag if tag else str(time.time()))
```
Comment (dimaspivak) [inline ambariclient/models.py]: `tag or str(time.time())` should work

File: ambariclient/models.py (hunk 2 above)
Comment (srids) [reply ambariclient/models.py]: gave few comments before and added a docstring with some more detials

**Pull request merged**